source/opt/local_single_store_elim_pass.cpp

```diff
@@ -319,6 +120,127 @@ void LocalSingleStoreElimPass::InitExtensions() {
       "SPV_EXT_descriptor_indexing",
   });
 }
+bool LocalSingleStoreElimPass::ProcessVariable(ir::Instruction* var_inst) {
+  vector<ir::Instruction*> users;
+  FindUses(var_inst, &users);
+
+  ir::Instruction* store_inst = FindSingleStoreAndCheckUses(var_inst, users);
+
+  if (store_inst == nullptr) {
+    return false;
+  }
+
+  return RewriteLoads(store_inst, users);
+}
+
+ir::Instruction* LocalSingleStoreElimPass::FindSingleStoreAndCheckUses(
+    ir::Instruction* var_inst, const vector<ir::Instruction*>& users) const {
+  // Make sure there is exactly 1 store.
+  ir::Instruction* store_inst = nullptr;
+
+  // If |var_inst| has an initializer, then that will count as a store.
+  if (var_inst->NumInOperands() > 1) {
+    store_inst = var_inst;
+  }
+
+  for (ir::Instruction* user : users) {
+    switch (user->opcode()) {
+      case SpvOpStore:
+        if (store_inst == nullptr) {
+          store_inst = user;
+        } else {
+          // More than 1 store.
+          store_inst = nullptr;
+          break;
+        }
+        break;
+      case SpvOpAccessChain:
+      case SpvOpInBoundsAccessChain:
+        if (FeedsAStore(user)) {
+          // Has a partial store.  Cannot propagate that.
+          store_inst = nullptr;
+          break;
+        }
+        break;
+      case SpvOpLoad:
+      case SpvOpImageTexelPointer:
```

Comment:
SpvOpImageTexelPointer can feed a "write" operation, right? Probably either need to verify or have this disqualify the variable to be safe, right?

Reply:
No I do not want it to disqualify the symbol.  It cannot feed and OpImageWrite, but it can feed atomic operations that write to the memory location.  However, I think this goes back to the issue we discussed before about the handle model.  When you do an OpImageWrite, what memory are you actually writing to?

```
   %gTexture = OpVariable %_ptr_UniformConstant_7 UniformConstant                                                                  
...
        %105 = OpLoad %7 %gTexture
               OpImageWrite %105 %int_0 %float_0
```

The write cannot be to %105 it is an ssa symbol.  It cannot be to the memory pointed to by %gTexture because it is in UniformConstant.  It write to the "image memory" that is access by looking at the data that %gTexture points to (the stuff in the UniformConstant memory.).

It is the same thing for OpImageTexelPointer.  The result of that instruction does not point to the function scope memory, but something else.  Whatever happens after to the result of the OpImageTexelPointer instruction, it cannot change the memory in the function scope variable.

Reply:
Agreed.